src/feature/hs/hs_dos.c

```diff
@@ -0,0 +1,45 @@
+/* Copyright (c) 2019, The Tor Project, Inc. */
+/* See LICENSE for licensing information */
+
+/**
+ * \file hs_dos.c
+ **/
+
+#define HS_DOS_PRIVATE
+
+#include "core/or/circuitlist.h"
+
+#include "hs_dos.h"
+
+/*
+ * Public API.
+ */
+
+/* Return true iff an INTRODUCE2 cell can be sent on the given service
+ * introduction circuit. */
+bool
+hs_dos_can_send_intro2(or_circuit_t *s_intro_circ)
+{
+  tor_assert(s_intro_circ);
+  tor_assert(TO_CIRCUIT(s_intro_circ)->purpose ==
+             CIRCUIT_PURPOSE_INTRO_POINT);
+
+  /* This is called just after we got a valid and parsed INTRODUCE1 cell. The
+   * service has been found and we have its introduction circuit.
+   *
+   * First, the INTRODUCE2 bucket will be refilled (if any). Then, decremented
```

Comment:
This might be a comment towards the token bucket API, but this unconditional refill seems a bit weird in a function that is supposed to decrement stuff. As it seems it's a conditional refill, but this is not obvious from the function name. We can either rename the token bucket function, or maybe let's clarify the cryptic "(if any)") part of the comment above.

Reply:
Probably want a new ticket for this.